### PR TITLE
Numpy add numpy op indices

### PR DESCRIPTION
**Pull request by gyshi**
## Description ##
(Brief description on what this PR is about)

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: src/operator/numpy/np_init_op.h [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+      MXNET_ADD_ALL_TYPES
+      .describe("Target data type.");
+    DMLC_DECLARE_FIELD(ctx)
+    .set_default("")
+    .describe("Context of output, in format [cpu|gpu|cpu_pinned](n)."
+              "Only used for imperative calls.");
+  }
+};
+
+inline bool NumpyIndicesShape(const nnvm::NodeAttrs& attrs,
```
Comment (haojin2) [inline src/operator/numpy/np_init_op.h]: put this function in the `.cc` file.

File: src/operator/numpy/np_init_op.h [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+                              mxnet::ShapeVector* in_shapes,
+                              mxnet::ShapeVector* out_shapes) {
+  const IndicesOpParam& param = nnvm::get<IndicesOpParam>(attrs.parsed);
+  CHECK_EQ(in_shapes->size(), 0U);
+  CHECK_EQ(out_shapes->size(), 1U);
+  CHECK_GE(param.dimensions.ndim(), 0)
+    << "_npi_indices dimensions the number of dim must not be less than  0";
```
Comment (haojin2) [inline src/operator/numpy/np_init_op.h]: remove extra space before `0`

Comment (haojin2) [reply src/operator/numpy/np_init_op.h]: "the number of dims of dimensions param must be non-negative"

File: src/operator/numpy/np_init_op.h [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+  mxnet::TShape param_dim = param.dimensions;
+  if (!shape_is_known(param_dim)) return false;
+  for (int i = 0; i < param_dim.ndim(); ++i) {
+    CHECK_GE(param_dim[i], 0) << "_npi_indices dimensions must be non-negative";
```
Comment (haojin2) [inline src/operator/numpy/np_init_op.h]: "all sizes of dims of dimensions params must be non-negative"

Comment (haojin2) [reply src/operator/numpy/np_init_op.h]: Also I think `shape_is_known(param_dim)` is already doing the same checks: https://github.com/apache/incubator-mxnet/blob/master/include/mxnet/tuple.h#L642-L662

Comment (gyshi) [reply src/operator/numpy/np_init_op.h]: thx, i have resolved.

**Comment by gyshi** [comment]
> backward compatibility check is failing for this PR:
> Can you check this and make sure that it passes - http://jenkins.mxnet-ci.amazon-ml.com/blue/organizations/jenkins/restricted-backwards-compatibility-checker/detail/restricted-backwards-compatibility-checker/778/pipeline

i will fix it, i will add backward for indices

**Comment by haojin2** [comment]
@Vikas-kum Please see the actual error message on the MBCC error: `OSError: libtvm_runtime.so: cannot open shared object file: No such file or directory` This PR is not related to the failure, it may just happens to be the last PR merged on that day.

**Pull request merged**